accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1179,6 +1179,119 @@ all members of `V1, .. Vk` that are not redeclared by a declaration in _DV_
+
+We introduce an interface type `Any` which declares `operator ==`,
+`hashCode`, `runtimeType`, `noSuchMethod`, `toString` with the same
+signatures as `Object` does today, and `Object` as well as `Null` have
+`Any` as a direct superinterface.
+
+`Any` is sealed. *That is, we cannot create additional direct subtypes of
+`Any` This is required for the soundness of the rule below about `Object?`
+being a supertype of `Any`.*
+
+*This basically means that we eliminate the anomaly which is created by
+saying that `Object` declares those members, but `Null` "has them too,
+by magic". It also implies that all paths from the top type via `Object`
+are now one step longer, which is significant during the computation of the
+standard upper bound.*
+
+For subtyping, we specify that every type is a subtype of `void`, of
+`dynamic`, and of `Any`.
+
+Moreover, we specify that `Any` is a subtype of `Object?`. *This implies
+that we collapse `Any` and `Object?`, which is basically the same thing as
+promising that `Any` will never have other subtypes than `Null` and
+`Object`. Other types (type variables and extension types) will always be
+populated by objects which are of type `Object` or `Null`, which means that
+it is sound to claim that `Object?` and `Any` are "the same type" (mutual
+subtypes).*
+
+We specify that `NonNull(Any)` is `Object`. *This is sound, and it is
+needed in order to allow `Any` to be used as a more concise way to write
+`Object?`, which is presumably quite valuable.*
+
+
+### Changes to functions
+
+Let `f` be a function with declared return type `R` whose body is `async`.
+
+An existing rule states that it is a compile-time error if `R` is not a
+supertype of `Future<T>` for any `T`.
+
+This is changed as follows: If `R` is an extension type that implements
+`Future<R1>` for some `R1` which is also the representation type of `R`
+then let `S` be `Future<R1>`. Otherwise let `S` be `R`. Next, it is a
+compile-time error if `S` is not a supertype of `Future<T>` for any `T`.
+
+*In short, nothing changes, except that an `async` function can now return
+an extension type if that extension type declares that it "is a future",
+and it has a representation type which is that same future type (not, for
+example, a subtype like `MyFuture`).*
+
+We say that said `Future<T>` is _the future return type_ of `f`. The
+run-time type of the object returned by `f` is a type that implements its
+future return type.
+
+*This is sufficient to ensure that it is a sound return value both in the
+case where `f` returns an extension type and a non-extension type.*
+
+Similarly, let `f` be a function with return type `R` whose body is marked
+`sync*`. 
+
+If `R` is an extension type that implements `Iterable<R1>` for some `R1`
+which is also the representation type of `R` then let `S` be
+`Iterable<R1>`. Otherwise let `S` be `R`. Next, it is a compile-time error
+if `S` is not a supertype of `Iterable<T>` for any `T`.
```

Comment:
Same issue. The implemented type is a red herring, it cannot be used unless it precisely matches the representation type, so always use the representation type.

Here:
> * *iterableElementType*(`T?`) = *iterableElementType*(`T`)
> * *iterableElementType*(`FutureOr<T>`) = *iterableElementType*(`T`)
> * *iterableElementType*(`Iterable<R>`) = `R`
> * *iterableElementType*(`E`) = *iterableElementType*(`R`) if `E` is an extension type with representation `R`
> * *iterableElementType*(`dynamic`) = `dynamic`
> * *iterableElementType*(`void`) = `void`
> * *iterableElementType*('T') = `Object?` if `Iterable<Never>` \<: `T` (aka **TOP**(`T`) or **OBJECT**(`T`)).
> * *iterableElementType*('T') = &bot;

Reply:
> The implemented type is a red herring, it cannot be used unless it precisely matches the representation type

Which is indeed guaranteed with the rules as proposed.

Reply:
Missed the "which is also the representation type of `R`".

Saying that one type "is" also another type means we have to determine whether two type terms denote the same type.
If one is `Future<Object?>` and the other `Future<dynamic>`, then presumably this rule would not apply.

Also, i actualy *want* to allow 
```dart
extension type Delay._(Future<void> _delay) {
  Delay(Duration duration) => Delay._(Future.delayed(duration, (){}));
  Delay add(Duration duration) => Delay._(_delay.then((_) => Future.delayed(duration, () {})));
  Future<void> get whenDone => _delay;
  static Delay max(Iterable<Delay> delays) =>
      Delay._(Future.wait(delays as Iterable<Future<void>>).then((_) {})); 
  static Delay consecutive(Iterable<Delay> delays) async {
    for (var delay in delays) await delay._delay;
  }
}
```
which is not a future, but where the `consecutive` function is an `async` function which returns a `Delay`.

So, I really want *just* the representation type to count, because otherwise I have to introduce an indirection like:
```dart
  static Delay consecutive(Iterable<Delay> delays) => Delay._(() async {
    for (var delay in delays) await delay._delay;
  }());
```
and nobody will thank us for making them do that.
(I expect streams to be more likely operations to wrap as something else.)

Reply:
Closing, this text has been deleted.